src/ringmesh/io/geomodel_builder_resqml.cpp

```diff
@@ -0,0 +1,994 @@
+            else
+            {
+                region_index = 0;
+            }
+        }
+
+        return region_index;
+    }
+
+    bool GeoModelBuilderRESQMLImpl::read_volumes( const EpcDocument& pck )
+    {
+        std::vector< UnstructuredGridRepresentation* > unstructuredGridRepSet =
+            pck.getUnstructuredGridRepresentationSet();
+        if( unstructuredGridRepSet.empty() )
+        {
+            return true;
+        }
+
+        for( auto unstructured_grid : unstructuredGridRepSet )
+        {
+            showAllMetadata( unstructured_grid );
+
+            if( unstructured_grid->isPartial()
+                || !unstructured_grid->hasGeometry() )
+            {
+                Logger::err(
+                    "Attempted to read partial or empty UnstructuredGrid" );
+                continue;
+            }
+
+            auto mesh = VolumeMesh3D::create_mesh();
+            bool result = read_volume_rep( *mesh, *unstructured_grid );
+            ringmesh_assert( result );
+
+            // the volume mesh from resqml is here, need to find the
+            // corresponding region of in the GeoModel3D
+
+            const int region_index = find_matching_geomodel_region( *mesh );
+            if( region_index == -1 )
+            {
+                Logger::err( "I/O", "RESQML2 could not find a matching region "
+                                    "for the volumn mesh" );
+                return false;
+            }
+
+            // corresponding region found, build its volume mesh
+            const gmme_id region_id(
+                region_type_name_static(), (index_t) region_index );
+
+            auto mesh_builder =
+                builder_.geometry.create_region_builder( region_id.index() );
+
+            for( auto v : range( mesh->nb_vertices() ) )
+            {
+                mesh_builder->create_vertex( mesh->vertex( v ) );
+            }
+
+            for( auto cell : range( mesh->nb_cells() ) )
+            {
+                mesh_builder->create_cells(
+                    (index_t) 1, mesh->cell_type( cell ) );
+
+                const index_t nb_vertices = mesh->nb_cell_vertices( cell );
+                std::vector< index_t > cell_vertices( nb_vertices, 0 );
+                for( auto v : range( nb_vertices ) )
+                {
+                    ElementLocalVertex lv( cell, v );
+                    mesh_builder->set_cell_vertex(
+                        lv, mesh->cell_vertex( lv ) );
+                }
+            }
+
+            mesh_builder->connect_cells();
+
+            // property
+            const Region3D& cur_reg = geomodel_.region( region_id.index() );
+
+            for( auto prop_index :
+                range( unstructured_grid->getValuesPropertySet().size() ) )
+            {
+                AbstractValuesProperty* propVal =
+                    unstructured_grid->getValuesPropertySet()[prop_index];
+                // TODO: there is only one patch in a UnstructuredGrid
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS1135)